### PR TITLE
[Merged by Bors] - feat (NumberTheory/Harmonic): compute Gamma'(1/2)

### PR DESCRIPTION
**Pull request by loefflerd**
Show `Gamma'(1/2) = -√π * (γ + 2 * log 2)` and various other related results.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Harmonic/GammaDeriv.lean [hunk 1]
```diff
@@ -79,49 +81,151 @@ lemma deriv_Gamma_nat (n : ℕ) :
 
 
 lemma hasDerivAt_Gamma_nat (n : ℕ) :
-    HasDerivAt Gamma ((n)! * (-eulerMascheroniConstant + harmonic n)) (n + 1) :=
+    HasDerivAt Gamma ((n)! * (-γ + harmonic n)) (n + 1) :=
   (deriv_Gamma_nat n).symm ▸
     (differentiableAt_Gamma fun m ↦ (by linarith : (n : ℝ) + 1 ≠ -m)).hasDerivAt
 
-lemma eulerMascheroniConstant_eq_neg_deriv : eulerMascheroniConstant = -deriv Gamma 1 := by
+lemma γ_eq_neg_deriv : γ = -deriv Gamma 1 := by
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:88]: Can you add a deprecated alias for the old name? Just in case somebody has been using it...

Comment (loefflerd) [reply Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:88]: Oops, the change was unintentional, I'll revert it.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM!

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

File: Mathlib/NumberTheory/Harmonic/GammaDeriv.lean [hunk 2]
```diff
@@ -19,20 +19,22 @@ We prove the formula for the derivative of `Real.Gamma` at a positive integer:
...
 
+local notation "γ" => Real.eulerMascheroniConstant
+
 namespace Real
 
 /-- Explicit formula for the derivative of the Gamma function at positive integers, in terms of
 harmonic numbers and the Euler-Mascheroni constant `γ`. -/
 lemma deriv_Gamma_nat (n : ℕ) :
-    deriv Gamma (n + 1) = (n)! * (-eulerMascheroniConstant + harmonic n) := by
+    deriv Gamma (n + 1) = (n)! * (-γ + harmonic n) := by
```
Comment (sgouezel) [inline Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:29]:
```suggestion
    deriv Gamma (n + 1) = n ! * (-γ + harmonic n) := by
```
Same thing below, `n !` looks more readable to me than `(n)!`.

File: Mathlib/NumberTheory/Harmonic/GammaDeriv.lean [hunk 3]
```diff
@@ -79,49 +81,151 @@ lemma deriv_Gamma_nat (n : ℕ) :
...
-lemma hasDerivAt_Gamma_one : HasDerivAt Gamma (-eulerMascheroniConstant) 1 := by
+lemma hasDerivAt_Gamma_one : HasDerivAt Gamma (-γ) 1 := by
   simpa only [factorial_zero, cast_one, harmonic_zero, Rat.cast_zero, add_zero, mul_neg, one_mul,
     cast_zero, zero_add] using hasDerivAt_Gamma_nat 0
 
+lemma hasDerivAt_Gamma_one_half : HasDerivAt Gamma (-√π * (γ + 2 * log 2)) (1 / 2) := by
+  have h_diff {s : ℝ} (hs : 0 < s) : DifferentiableAt ℝ Gamma s :=
+    differentiableAt_Gamma fun m ↦ ((neg_nonpos.mpr m.cast_nonneg).trans_lt hs).ne'
+  have h_diff' {s : ℝ} (hs : 0 < s) : DifferentiableAt ℝ (fun s ↦ Gamma (2 * s)) s := by
+    exact DifferentiableAt.comp (g := Gamma) _ (h_diff <| mul_pos two_pos hs)
```
Comment (sgouezel) [inline Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:100]:
```suggestion
  have h_diff' {s : ℝ} (hs : 0 < s) : DifferentiableAt ℝ (fun s ↦ Gamma (2 * s)) s :
    DifferentiableAt.comp (g := Gamma) _ (h_diff <| mul_pos two_pos hs)
```
No need for `by exact`.

File: Mathlib/NumberTheory/Harmonic/GammaDeriv.lean [hunk 4]
```diff
@@ -79,49 +81,151 @@ lemma deriv_Gamma_nat (n : ℕ) :
...
+    · apply h_diff; norm_num -- s = 1/2
+    · exact ((h_diff (by norm_num)).hasDerivAt.comp_add_const).differentiableAt -- s = 1
+  _ = (deriv (fun s ↦ Gamma (2 * s) * 2 ^ (1 - 2 * s) * √π) (1 / 2)) + √π * γ := by
+    rw [funext Gamma_mul_Gamma_add_half]
+  _ = √π * (deriv (fun s ↦ Gamma (2 * s) * 2 ^ (1 - 2 * s)) (1 / 2) + γ) := by
+    rw [mul_comm √π, mul_comm √π, deriv_mul_const, add_mul]
+    · apply DifferentiableAt.mul
+      · exact DifferentiableAt.comp (g := Gamma) _ (by apply h_diff; norm_num) -- s = 1
+          (differentiableAt_id.const_mul _)
+      · exact (differentiableAt_const _).rpow (by fun_prop) two_ne_zero
```
Comment (sgouezel) [inline Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:121]:
```suggestion
    apply DifferentiableAt.mul
    · exact DifferentiableAt.comp (g := Gamma) _ (by apply h_diff; norm_num) -- s = 1
        (differentiableAt_id.const_mul _)
    · exact (differentiableAt_const _).rpow (by fun_prop) two_ne_zero
```
There is just one goal, so no need for focusing dot.

File: Mathlib/NumberTheory/Harmonic/GammaDeriv.lean [hunk 5]
```diff
@@ -79,49 +81,151 @@ lemma deriv_Gamma_nat (n : ℕ) :
...
+open scoped Real
+
+private lemma HasDerivAt.complex_of_real {f : ℂ → ℂ} {g : ℝ → ℝ} {g' s : ℝ}
+    (hf : DifferentiableAt ℂ f s) (hg : HasDerivAt g g' s) (hfg : ∀ s : ℝ, f ↑s = ↑(g s)) :
+    HasDerivAt f ↑g' s := by
+  refine HasDerivAt.congr_deriv hf.hasDerivAt ?_
+  rw [← (funext hfg ▸ hf.hasDerivAt.comp_ofReal.deriv :)]
+  exact hg.ofReal_comp.deriv
+
+private lemma differentiable_at_Gamma_nat_add_one (n : ℕ) :
```
Comment (sgouezel) [inline Mathlib/NumberTheory/Harmonic/GammaDeriv.lean:161]: I wouldn't make this one private, it might be handy.

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

**Comment by loefflerd** [comment]
Thanks Michael and Sebastien for the reviews!

bors r+

**Pull request closed without merging**